### PR TITLE
Support TypeScript in extensions

### PR DESCRIPTION
**Pull request by choumx**
First PR for #13791.

- New dev-dependencies: `typescript` and `tsickle`
- Support `typescript` option for extension declarations in gulpfile.js
- Extensions that specify this option will have all TS files reachable by its entry point transpiled to Closure-annotated JS before compilation (and deleted afterwards)

Note: Requires angular/tsickle#758 to work properly.

/to @rsimha @kristoferbaxter

### CODE REVIEW

File: gulpfile.js [hunk 1]
```diff
@@ -64,114 +68,112 @@ const unminifiedRuntimeTarget = 'dist/amp.js';
 const unminified3pTarget = 'dist.3p/current/integration.js';
 
 // Each extension and version must be listed individually here.
-declareExtension('amp-3q-player', '0.1', {hasCss: false});
```
Comment (kristoferbaxter) [inline gulpfile.js]: Removing the option for `hasCss` seems disconnected from the goal of adding TypeScript support. Are these changes necessary or can we ensure the arity of the declareExtension function is always matching?

File: gulpfile.js [hunk 2]
```diff
@@ -538,21 +538,19 @@ function buildExtension(name, version, hasCss, options, opt_extraGlobs) {
...
-      }
-      return buildExtensionJs(path, name, version, options);
     });
-  } else {
-    return buildExtensionJs(path, name, version, options);
+    if (options.compileOnlyCss) {
+      return promise;
+    }
   }
+  return promise.then(() => buildExtensionJs(path, name, version, options));
```
Comment (kristoferbaxter) [inline gulpfile.js]: Nice clear up here. Good for @erwinmombay and @rsimha to take a look as well.

File: gulpfile.js [hunk 3]
```diff
@@ -600,14 +598,10 @@ function buildExtensionJs(path, name, version, options) {
     throw new Error('Unsupported loadPriority: ' + options.loadPriority);
   }
   const priority = options.loadPriority ? 'p:"high",' : '';
-  return compileJs(path + '/', filename, './dist/v0', {
-    watch: options.watch,
-    preventRemoveAndMakeDir: options.preventRemoveAndMakeDir,
-    minify: options.minify,
+  return compileJs(path + '/', filename, './dist/v0', Object.assign(options, {
     toName: name + '-' + version + '.max.js',
```
Comment (kristoferbaxter) [inline gulpfile.js]: If you're moving this over to Object.assign perhaps it's also worth reformatting these to template literals.

```javascript
Object.assign(options, {
  toName: `${name}-${version}.max.js`,
  minifiedName: `${name}-{version}.js`,
  latestName: `${name}-latest.js`,
  wrapper: options.noWrapper ? '' : `(self.AMP=self.AMP||[]).push({n:"${name}",${priority}v:"${internalRuntimeVersion}",f:(function(AMP){<%= contents %>\n})});`
})
```

File: gulpfile.js [hunk 4]
```diff
@@ -990,22 +986,35 @@ function concatFilesToString(files) {
  */
 function compileJs(srcDir, srcFilename, destDir, options) {
   options = options || {};
+
+  const entryPoint = path.join(srcDir, srcFilename);
+
+  // Transpile TS to Closure-annotated JS before actual bundling or compile.
+  if (options.typeScript) {
```
Comment (kristoferbaxter) [inline gulpfile.js]: Sad to see the auto-discovery of typescript components go away with this change and move to explicit options directive.

What's the reason?

File: gulpfile.js [hunk 5]
```diff
@@ -1063,10 +1072,16 @@ function compileJs(srcDir, srcFilename, destDir, options) {
             .pipe($$.rename(destFilename))
             .pipe(lazywrite())
             .on('end', function() {
-              appendToCompiledFile(srcFilename, destDir + '/' + destFilename);
+              appendToCompiledFile(srcFilename,
+                  path.join(destDir, destFilename));
             }))
         .then(() => {
```
Comment (kristoferbaxter) [inline gulpfile.js]: Cannot wait until we move to Node 8+ as a requirement so these can become async/await.

File: gulpfile.js [hunk 6]
```diff
@@ -1106,6 +1121,67 @@ function compileJs(srcDir, srcFilename, destDir, options) {
...
 }
 
+/**
+ * Given a file path `foo/bar.js`, transpiles the TypeScript entry point of
+ * the same name `foo/bar.ts` and all direct and indirect TypeScript imports.
+ *
+ * @param {string} srcDir
+ * @param {string} srcFilename
+ */
+function transpileTs(srcDir, srcFilename) {
```
Comment (kristoferbaxter) [inline gulpfile.js]: Should this function live in gulpfile? Or be imported from standalone location?

File: gulpfile.js [hunk 7]
```diff
@@ -1106,6 +1121,67 @@ function compileJs(srcDir, srcFilename, destDir, options) {
...
+  const startTime = Date.now();
+
+  const tsEntry = path.join(srcDir, srcFilename).replace('.js', '.ts');
+  const tsConfig = ts.convertCompilerOptionsFromJson({
```
Comment (kristoferbaxter) [inline gulpfile.js]: Does this imply a specific convention for the location of a `tsconfig.json` file relative to the extension?

File: gulpfile.js [hunk 8]
```diff
@@ -1106,6 +1121,67 @@ function compileJs(srcDir, srcFilename, destDir, options) {
...
+  const compilerHost = ts.createCompilerHost(tsOptions);
+  const program = ts.createProgram([tsEntry], tsOptions, compilerHost);
+
+  const transformerHost = {
+    shouldSkipTsickleProcessing: () => false,
+    transformTypesToClosure: true,
+    options: tsOptions,
+    host: compilerHost,
+  };
+  const emitResult = tsickle.emitWithTsickle(program, transformerHost,
```
Comment (kristoferbaxter) [inline gulpfile.js]: For debugging purposes it might be beneficial to be able to disable tsickle and use the standard TypeScript compiler directly.

File: gulpfile.js (hunk 6 above)
Comment (choumx) [reply gulpfile.js]: Good point, this file is huge. Done.

File: gulpfile.js (hunk 7 above)
Comment (kristoferbaxter) [reply gulpfile.js]: Makes sense in most cases. However, for TS there are different configurations for code in a WorkerGlobalScope versus the UI thread. This might cause a few issues in those scenarios.

File: gulpfile.js (hunk 8 above)
Comment (kristoferbaxter) [reply gulpfile.js]: Good with me. Awesome work.

File: gulpfile.js (hunk 4 above)
Comment (kristoferbaxter) [reply gulpfile.js]: Makes sense. Happy to go this route and minimize differences.

File: gulpfile.js (hunk 1 above)
Comment (kristoferbaxter) [reply gulpfile.js]: This code shouldn't be hot path, but we should verify the impact of mismatching arity.

This isn't a blocker by any means, just good for us to know.

File: gulpfile.js (hunk 2 above)
Comment (jridgewell) [inline gulpfile.js]: This will end up compiling JS on css-only paths.

File: gulpfile.js [hunk 9]
```diff
@@ -1063,10 +1073,16 @@ function compileJs(srcDir, srcFilename, destDir, options) {
...
-              appendToCompiledFile(srcFilename, destDir + '/' + destFilename);
+              appendToCompiledFile(srcFilename,
+                  path.join(destDir, destFilename));
             }))
         .then(() => {
           endBuildStep('Compiled', srcFilename, startTime);
+
+          // Remove intemediary, transpiled JS files after compilation.
+          if (options.typeScript) {
+            removeJsFilesInDirectory(srcDir);
```
Comment (jridgewell) [inline gulpfile.js]:
```js
rimraf.sync(path.join(srcdir, '**/*.js'))

File: build-system/typescript.js [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+
+/**
+ * Given a file path `foo/bar.js`, transpiles the TypeScript entry point of
+ * the same name `foo/bar.ts` and all direct and indirect TypeScript imports.
+ *
+ * @param {string} srcDir
+ * @param {string} srcFilename
+ */
+exports.transpileTs = function(srcDir, srcFilename) {
+  const tsEntry = path.join(srcDir, srcFilename).replace('.js', '.ts');
```
Comment (jridgewell) [inline build-system/typescript.js]: Might be better to update `buildExtensionJs` to use `.ts` when typescript is set.

File: gulpfile.js (hunk 2 above)
Comment (choumx) [reply gulpfile.js]: As discussed, this is handled above in this function.

File: build-system/typescript.js (hunk 10 above)
Comment (choumx) [reply build-system/typescript.js]: I did that at first but we still actually want to compile the `.js` file normally after the TS transpile. This does look a bit ugly though we'll clean it up as we expand TS support.

**Comment by choumx** [comment]
Note: This PR requires https://github.com/angular/tsickle/pull/758 to work properly.

File: gulpfile.js (hunk 9 above)
Comment (choumx) [reply gulpfile.js]: Done.

File: gulpfile.js (hunk 5 above)
Comment (rsimha) [reply gulpfile.js]: @kristoferbaxter We currently require node 6+ because it is the lowest Active LTS version.  This will change in a month, when node 8+ becomes the lowest Active LTS version (see https://github.com/nodejs/Release), so I support making Node 8+ a requirement soon.

cc @jridgewell for comment.

File: build-system/typescript.js [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+  if (tsConfig.errors.length) {
+    log(colors.red('TSickle:'), tsickle.formatDiagnostics(tsConfig.errors));
+  }
+
+  const compilerHost = ts.createCompilerHost(tsOptions);
+  const program = ts.createProgram([tsEntry], tsOptions, compilerHost);
+
+  // TODO(choumx): This was partially copy-pasta'd from tsickle. Add a default
+  // to tsickle so this can be an optional param.
+  const pathToModuleName = (context, fileName) => {
```
Comment (rsimha) [inline build-system/typescript.js]: I wonder if `path.resolve()` can trivially do what this function is doing: https://nodejs.org/api/path.html#path_path_resolve_paths

File: gulpfile.js [hunk 12]
```diff
@@ -600,27 +598,21 @@ function buildExtensionJs(path, name, version, options) {
     throw new Error('Unsupported loadPriority: ' + options.loadPriority);
   }
   const priority = options.loadPriority ? 'p:"high",' : '';
-  return compileJs(path + '/', filename, './dist/v0', {
-    watch: options.watch,
```
Comment (rsimha) [inline gulpfile.js]: `compileJs` uses `options.watch` to determine whether to call `rebundle` or not. See https://github.com/ampproject/amphtml/blob/master/gulpfile.js#L1087=L1107

Will `gulp`, `gulp watch`, `gulp build`, and `gulp dist` all continue to work with this change? Also, will they work with `--noextensions` and `--extensions`?

File: gulpfile.js (hunk 5 above)
Comment (jridgewell) [reply gulpfile.js]: I think I lost the maintainance LTS debate, so moving to the latest active LTS is fine.

Comment (rsimha) [reply gulpfile.js]: PR sent for the bump up to node 8. #13848.

File: gulpfile.js (hunk 12 above)
Comment (choumx) [reply gulpfile.js]: `options.watch`, `options.minify` and others are retained due to `Object.assign`.

File: gulpfile.js (hunk 5 above)
Comment (rsimha) [reply gulpfile.js]: PR landed. Node 8 is now the minimum required version for AMP development.
